File: Assignments/Week 4/HomeworkEveryone/vertex.cpp
/**
 * @file vertex.cpp
 *
 * @author Majbrit Schoettner
 * Contact: mshoettner@stud.hs-bremen.de
 * @author Reeder Ward
 * Contact: rward@stud.hs-bremen.de
 * @author David Melamed
 * Contact: dmelamed@stud.hs-bremen.de
 */

//NEW
#include "vertex.h"
#include <iostream>
using namespace std;

Vertex::Vertex()
{

}

Vertex::Vertex(float x, float y, float z)
{
    xyz[0] = x;
    xyz[1] = y;
    xyz[2] = z;
}

Vertex::Vertex( float point[3])
{
    xyz[0] = point[0];
    xyz[1] = point[1];
    xyz[2] = point[2];
}

void Vertex::print(string description )
{
    cout << description << "[ " << xyz[0] << "; " << xyz[1] << "; " << xyz[2] << " ]" << endl;
}

Vertex operator%(Vertex a, Vertex b)
{
    Vertex c;
    c.xyz[0] = a.xyz[1]*b.xyz[2] - a.xyz[2]*b.xyz[1];
    c.xyz[1] = a.xyz[2]*b.xyz[0] - a.xyz[0]*b.xyz[2];
    c.xyz[2] = a.xyz[0]*b.xyz[1] - a.xyz[1]*b.xyz[0];

    return c;
}

Vertex operator-(Vertex a, Vertex b)
{
    Vertex c;
    c.xyz[0] = a.xyz[0] - b.xyz[0];
    c.xyz[1] = a.xyz[1] - b.xyz[1];
    c.xyz[2] = a.xyz[2] - b.xyz[2];

    return c;
}

File: Assignments/Project 2/checkIn/vertex.py
__author__ = "David Melamend, Reeder Ward, Majbrit Schoettner"
__email__ = "dmelamed@stud.hs-bremen.de, rward@stud.hs-bremen.de, mshoettner@stud.hs-bremen.de"
__title__ = "Implement Toolchain for .OBJ Objects in PyQt5 GUI to Create a 3D Modelling Software"
__course__ = "Computer Graphics"

class vertex:    
    def __init__(self, x, y, z):
        self.xyz = [float(x), float(y), float(z)]
        #print("vertex created")

    # print coordinates
    def printXYZ(self):
        print("xyz = [ ", self.xyz[0],"; ", self.xyz[1],"; ", self.xyz[2]," ]")

    # +
    def __add__(self, other):
        return vertex(self.xyz[0] + other.xyz[0], self.xyz[1] + other.xyz[1], self.xyz[2] + other.xyz[2])

    # -
    def __sub__(self, other):
        return vertex(self.xyz[0] - other.xyz[0], self.xyz[1] - other.xyz[1], self.xyz[2] - other.xyz[2])

    # +=
    def __iadd__(self, other):
        self.xyz[0] = self.xyz[0] + other.xyz[0]
        self.xyz[1] = self.xyz[1] + other.xyz[1]
        self.xyz[2] = self.xyz[2] + other.xyz[2]
        return self

    # -=
    def __isub__(self, other):
        self.xyz[0] = self.xyz[0] - other.xyz[0]
        self.xyz[1] = self.xyz[1] - other.xyz[1]
        self.xyz[2] = self.xyz[2] - other.xyz[2]
        return self

    # *=
    def __imul__(self, other):
        self.xyz[0] = self.xyz[0] * other
        self.xyz[1] = self.xyz[1] * other
        self.xyz[2] = self.xyz[2] * other
        return self

    # %
    def __mod__(self, other):
        return vertex(self.xyz[1]*other.xyz[2] - self.xyz[2]*other.xyz[1], self.xyz[2]*other.xyz[0] - self.xyz[0]*other.xyz[2], self.xyz[0]*other.xyz[1] - self.xyz[1]*other.xyz[0])

    # /
    def __truediv__(self, other):
        return vertex(self.xyz[0]/other, self.xyz[1]/other, self.xyz[2]/other)

    # * 
    def __mul__(self, other):
        if type(other) == vertex:
            result = self.xyz[0] * other.xyz[0] + self.xyz[1] * other.xyz[1] + self.xyz[2] * other.xyz[2]
        else:
            result =vertex(other*self.xyz[0], other*self.xyz[1], other*self.xyz[2])
        return result
    def __rmul__(self, other):
        return vertex.__mul__(self,other) 



